Dora Fine-tuning added to examples

shirinyamani opened the pull request:
In this PR a dir named "dora finetuning" is added to the examples dir. The content of it includes a notebook teaching how to use Dora technique along with an script of fine-tuning using Dora. Hope this helps the community!

BenjaminBossan reviewed (changes requested):
Thanks a lot for adding the notebook, script, and description for DoRA training. This looks pretty nice and is a good addition for PEFT!

I could run the script and notebook in general, but I stumbled onto some smaller issues. Please check my comments.

Also, I saw that the notebook and script are similar but there are also differences. Just as an example, the script uses `tokenize_function`. Is there a specific reason for this discrepancy? If yes, it would make sense to explain it (e.g. use the script for X and the notebook for Y), otherwise, I'd suggest to align the two.

Finally, could you please run `make style` on the PR?

shirinyamani commented on examples/dora_finetuning/dora_finetuning.py:
```diff
@@ -0,0 +1,185 @@
+    parser.add_argument("--data_path", type=str, default="timdettmers/openassistant-guanaco", help="Dataset path or name")
+    parser.add_argument("--output_dir", type=str, default="path/to/output", help="Output directory for the fine-tuned model")
+    parser.add_argument("--batch_size", type=int, default=1, help="Batch size")
+    parser.add_argument("--use_compile", default=False, help="Use Compile")
```
I  removed use_compile from both the .py and the readme files, I ran into same index error, I got surprised since according to this [torch.compile](https://huggingface.co/docs/peft/main/en/developer_guides/torch_compile) document on HF torch.compile() should perfectly work for the Lora + Dora adapter setup.  🤔
For now, use_compile is completely removed but I am wondering to kindly know your idea on "why" they do not work well together?

shirinyamani commented on examples/dora_finetuning/README.md:
```diff
@@ -0,0 +1,106 @@
+    --learning_rate 3e-4 \
+    --cutoff_len 512 \
+    --val_set_size 500 \
+    --use_compile True \
```
I  removed use_compile from both the dora_finetuning.py and the readme files, so this should perfectly work now!

shirinyamani commented on examples/dora_finetuning/dora_finetuning.py:
```diff
@@ -0,0 +1,185 @@
+            token=hf_token, 
+            quantization_config=BitsAndBytesConfig(
+                load_in_4bit=True, 
+                bnb_4bit_compute_dtype=torch.bfloat16,
```
addressed to support for older GPUs 
`torch.bfloat16 if torch.cuda.is_available() and torch.cuda.is_bf16_supported() else torch.float16`

shirinyamani commented on examples/dora_finetuning/dora_finetuning.py:
```diff
@@ -0,0 +1,185 @@
+    # if use_peft is True THEN, DORA setup
+    if use_peft: #TODO: add other config than lora  peft_configs = (LoraConfig, AdaptionPromptConfig, PrefixTuningConfig)
+        lora_config = LoraConfig(
+            use_dora=True, 
```
I input the use_dora as you kindly recommended, so what you are mentioning could have implemented in 2 ways;
1. (more user-friendly): Remove `use_peft` and instead add use_dora as you recommended. this way the default setup would be Lora then the user get the freedom of choice on inputting the `use_dora` to see the change (implemented in updated PR), the downside is, the Lora is set as default and user do not get to choose to cancel it as before (use_peft). But I believe that is the whole point bc user wanna see the difference between 2 setups!

2. (might be confusing) the user is given the freedom of chose first to use_peft then if yes, now choose which config (Lora, prefix, etc -configs) then on top of this if Dora is supported on the desired setup then set it to True!
tbh, I found it a bit confusing for the end user!

BenjaminBossan commented on examples/dora_finetuning/dora_finetuning.py:
```diff
@@ -0,0 +1,173 @@
+    #Load model in Quantized and Peft configz  
+    #QDoRA: IF YOU WANNA QUANTIZE THE MODEL
+    if quantize:
+        model = AutoModelForCausalLM.from_pretrained(
```
When quantizing, please call `prepare_model_for_kbit_training`, same as in the notebook.

shirinyamani replied:
updated to reflect `model = prepare_model_for_kbit_training(model, use_gradient_checkpointing=True)`

BenjaminBossan reviewed (changes requested):
Thanks for all the changes, this PR is almost ready to be merged. I found a few minor things that need yet to be changed, please check my comments.

After the changes to the notebook, did you re-run it? If not, could you please do that.

Furthermore, could you please run `make style`? Otherwise, CI won't pass.

> I got surprised since according to this [torch.compile](https://huggingface.co/docs/peft/main/en/developer_guides/torch_compile) document on HF torch.compile() should perfectly work for the Lora + Dora adapter setup. So, would it mean that the changes (according to your provided example) should be applied to the tokenization?

Honestly, I'm not 100% sure what's going on but there is an issue with max token length when using `torch.compile`, which requires extra care with handling tokenization, as shown in the test that I linked. This issue is not directly PEFT related.

shirinyamani commented on examples/dora_finetuning/README.md:
````diff
@@ -0,0 +1,106 @@
+
+
+Run the finetuning script simply by running:
+```python
````
fixed

shirinyamani commented:
> Thanks for all the changes, this PR is almost ready to be merged. I found a few minor things that need yet to be changed, please check my comments.
> 
> After the changes to the notebook, did you re-run it? If not, could you please do that.
>  Yes, I just did and all is perfectly working! The newly-ran notebook is added!

> Furthermore, could you please run `make style`? Otherwise, CI won't pass.
> Just ran successfully! Here is the output;
`(llm) [yamani.shirin@mg6 peft]$ make style
ruff check --fix src tests examples docs scripts docker
All checks passed!
ruff format src tests examples docs scripts docker
165 files left unchanged
doc-builder style src/peft tests docs/source --max_len 119`

> > I got surprised since according to this [torch.compile](https://huggingface.co/docs/peft/main/en/developer_guides/torch_compile) document on HF torch.compile() should perfectly work for the Lora + Dora adapter setup. So, would it mean that the changes (according to your provided example) should be applied to the tokenization?
> 
> Honestly, I'm not 100% sure what's going on but there is an issue with max token length when using `torch.compile`, which requires extra care with handling tokenization, as shown in the test that I linked. This issue is not directly PEFT related.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_1885). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

BenjaminBossan commented:
Thanks for the updates. Now we only need to make the code style checks pass and we should be good.

So something must have gone wrong with `make style`. Let's solve this step by step:

- ensure that you're in the project root (not in the examples directory)
- ensure that the correct Python virtual env is active
- check that the ruff version is correct: `ruff --version` => 0.4.10
- run `make style`
- there should be changes now, which should be committed and pushed to the PR.

If `make style` doesn't work for some reason, you can also call the commands explicitly:

```sh
ruff check --fix examples
ruff format examples
```

shirinyamani commented:
> Thanks for the updates. Now we only need to make the code style checks pass and we should be good.
> 
> So something must have gone wrong with `make style`. Let's solve this step by step:
> 
> * ensure that you're in the project root (not in the examples directory)
> * ensure that the correct Python virtual env is active
> * check that the ruff version is correct: `ruff --version` => 0.4.10
> * run `make style`
> * there should be changes now, which should be committed and pushed to the PR.
> 
> If `make style` doesn't work for some reason, you can also call the commands explicitly:
> 
> ```shell
> ruff check --fix examples
> ruff format examples
> ```

I am really grateful for your hint! I just pushed the changes after the ruff shell commands. I appreciate if you let me know it get passed!

BenjaminBossan reviewed (approved):
Thanks for fixing the coding style. That part of the CI is now passing. There are still some failing jobs, but those are unrelated to the PR. Therefore, it can now be merged. Thanks for providing this well written DoRA example.

BenjaminBossan commented:
For reference, this is the diff for the change:

```diff
modified   examples/dora_finetuning/dora_finetuning.py
@@ -6,7 +6,7 @@ from transformers import (
     AutoModelForCausalLM,
     AutoTokenizer,
     BitsAndBytesConfig,
-    DataCollatorWithPadding,
+    DataCollatorForLanguageModeling,
     Trainer,
     TrainingArguments,
 )
@@ -95,7 +95,7 @@ def train_model(
     tokenized_datasets = dataset.map(tokenize_function, batched=True, remove_columns=dataset["train"].column_names)
 
     # Data collator to dynamically pad the batched examples
-    data_collator = DataCollatorWithPadding(tokenizer)
+    data_collator = DataCollatorForLanguageModeling(tokenizer, mlm=False)
 
     # Define training arguments
     training_args = TrainingArguments(
```

shirinyamani commented:
I agree with you, just reviewed the code. Do I need to fork and update the changes ?  @BenjaminBossan

BenjaminBossan commented:
Thanks for checking. If you want to fix this, create a new PR based off the current `main` branch of PEFT. You don't need to create a new fork, as you [already have a PEFT fork](https://github.com/shirinyamani/peft/tree/main), you just need to refresh it so that it is up-to-date with PEFT main. There should be a button when you visit the link:

![image](https://github.com/user-attachments/assets/9beee1c1-739b-4f2b-b87c-cb92c0dd8e8e)

BenjaminBossan commented:
Did you have time to check this @shirinyamani? If you don't have time, let me know and I can work on a fix myself.

shirinyamani commented:
Just updated! @BenjaminBossan , my apologies for the delay, I was sick terribly sick for a week! Now updated with the `DataCollatorForLanguageModeling`
